### PR TITLE
YAML consolidation: `netplan get` (FR-702)

### PR DESCRIPTION
**Pull request by schopin-pro**
## Description

Port the `netplan get` command to use the internal YAML generator, originally written for the NM-to-netplan system.
In order to make this work, we had to add a notion of "dirtyness" to the input data, as the user would want their explicit settings displayed even if they match the default value.

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

### CODE REVIEW

**Comment by schopin-pro** [comment]
Draft as it depends on #250 and #251 , and would also benefit from #249 being merged

File: tests/generator/base.py [hunk 1]
```diff
@@ -274,6 +274,7 @@ def validate_generated_yaml(self, yaml_input):
                     input_yaml = yaml.safe_load('')
 
             # Read output of the YAML generator (if any)
+            output_yaml = None
```
Comment (slyon) [inline tests/generator/base.py:277]: This seems unrelated. It will be overridden in either the `if` or the `else` statement. It is best practice to have a variable clearly defined on top before we do any logic on it, so we can leave it in there.
But just to clarify: This has no other purpose than better readability/clarity of the code, right?

File: netplan/cli/commands/get.py [hunk 2]
```diff
@@ -41,27 +42,37 @@ def run(self):
         self.parse_args()
         self.run_command()
 
+    def dump_state(self, key, np_state, output_file):
+        if key == 'all':
+            np_state.dump_yaml(output_file=output_file)
+            return
+
+        if not key.startswith('network'):
```
Comment (slyon) [inline netplan/cli/commands/get.py:50]: Hmm... I remember having a reason for doing that in the specific way it was done before; I can remember the details, though... Would you mind keeping it at this, to not change behaviour?
```suggestion
        if not key.startswith('network.') and not key == 'network':
```

I think one problem could be that if somebody tries to `netplan get networkINVALID`, this should return an error/empty tree. OTOH.. I think that might actually be handled by the `dump_yaml_subtree` function, as it won't find the `networkINVALID` prefix.

**Review by slyon — COMMENTED** [review]
Thank you for another great PR, Simon!
The dirty tracking functionality is really nice and should help us cleaning up some parts of the test environment, too, where we're currently doing some "normalization" of default values.

I left some inline comments. But again, nothing critical.

File: tests/generator/base.py (hunk 1 above)
Comment (schopin-pro) [reply tests/generator/base.py:277]: Oh, no, even better than that: it has no purpose. It's an accidental leftover of the previous version of #251 with its unfortunate side effect on empty files. I'll get rid of it.

File: netplan/cli/commands/get.py (hunk 2 above)
Comment (schopin-pro) [reply netplan/cli/commands/get.py:50]: One reason would be the handling of documents with more than one top-level key, but since we go through the entire parsing process, those would get tossed out anyway.

Unless you give me an actual case where the new code is problematic, I'd rather keep it as is, as I've noticed that weird code with unclear rationales tend to calcify.

I've added a test case for the `networkINVALID` query, though :)

File: src/parse.c [hunk 3]
```diff
@@ -2228,128 +2276,128 @@ static const mapping_entry_handler dhcp6_overrides_handlers[] = {
...
+    {"ipv6-mtu", YAML_SCALAR_NODE, {.generic=handle_netdef_guint}, netdef_offset(ipv6_mtubytes)},  \
+    {"ipv6-privacy", YAML_SCALAR_NODE, {.generic=handle_netdef_bool}, netdef_offset(ip6_privacy)}, \
+    {"link-local", YAML_SEQUENCE_NODE, {.generic=handle_link_local}},                                    \
+    {"macaddress", YAML_SCALAR_NODE, {.generic=handle_netdef_mac}, netdef_offset(set_mac)},        \
+    {"mtu", YAML_SCALAR_NODE, {.generic=handle_netdef_guint}, netdef_offset(mtubytes)},            \
+    {"nameservers", YAML_MAPPING_NODE, {.map={FALSE, {.handlers=nameservers_handlers}}}},                           \
+    {"optional", YAML_SCALAR_NODE, {.generic=handle_netdef_bool}, netdef_offset(optional)},        \
+    {"optional-addresses", YAML_SEQUENCE_NODE, {.generic=handle_optional_addresses}},                    \
+    {"renderer", YAML_SCALAR_NODE, {.generic=handle_netdef_renderer}},                                   \
+    {"routes", YAML_SEQUENCE_NODE, {.generic=handle_routes}},                                            \
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/parse.c]: I ended up removing them here, as the fields are way too varying for it to render cleanly, but I've kept them in the other instance, since all fields have the same type, thus same approximate length.

File: src/parse.c [hunk 4]
```diff
@@ -187,16 +187,25 @@ netplan_netdef_new(NetplanParser *npp, const char* id, NetplanDefType type, Netp
...
-    /* handler for the value of this key */
-    node_handler handler;
-    /* if type == YAML_MAPPING_NODE and handler is NULL, use process_mapping()
-     * on this handler map as handler */
-    const struct mapping_entry_handler_s* map_handlers;
+    union {
+        node_handler generic;
+        custom_map_handler variable;
+        struct {
+            gboolean custom_handling;
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/parse.c]: Turns out it could be implied :)

**Comment by schopin-pro** [comment]
Tests run on my local LXD, as GH ACtions seems to be having troubles today:

```
autopkgtest [11:31:23]: @@@@@@@@@@@@@@@@@@@@ summary
ovs                  PASS
ethernets            PASS
bridges              PASS
bonds                PASS
routing              PASS
vlans                PASS
wifi                 PASS
tunnels              PASS
scenarios            PASS
regressions          PASS
autostart            PASS
cloud-init           PASS
```

File: netplan/cli/commands/get.py (hunk 2 above)
Comment (slyon) [reply netplan/cli/commands/get.py:50]: That's fair, I guess! All our tests are passing. Should we hit a regression, we'll fix it up and add a new test case for it.

**Review by slyon — APPROVED** [review]
Thank you for the V2 improvements and for running the integration tests locally!

LGTM :eyes:
